### PR TITLE
replace `hasmethod` by correct implementation

### LINKED ISSUE #25: @implement requires explicit argument types
The argument requires to specify an argument type for each argument - could default to Any

Yes, but I am unsure if Any is the right default. When I tried to use it with the AbstractArray interface, which uses duck typing in the interface, I figured that I need to use Base.Bottom instead. See the LinearIndexing trait example. If we want to impose a default then I think Bottom is more appropriate because otherwise the implementer must define the function that accepts explicitly Any.

### PR DESCRIPTION
**Pull request by KlausC**
Fixes #25.

This PR includes the previous #28.
The `Base.hasmethod` has a known bug (in highly optimized C-code, see JuliaLang/julia#30808), which is not being removed for a year. It does not work correctly, if the one of the method argument types is `Base.Bottom`.
As we don't need the high performance, a solution was implemented in `Julia`.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -173,6 +173,31 @@ module Interfaces
...
+    struct Penguin end
+    @trait Dive
+    @assign Penguin with Dive
+    @implement CanDive by dive1(::Integer)           # missing argument name
+    @implement CanDive by dive2(::Vector{<:Integer}) # parameterized type
+    if VERSION >= v"1.2"
+    @implement CanDive by dive31(x::Real;)            # keyword arguments
+    @implement CanDive by dive32(x::Real; kw::Real)   # keyword arguments
+    @implement CanDive by dive33(x::Real; kw1::Real, kw2) # keyword arguments
+    @implement CanDive by dive34(x;kw1)
```
Comment (tk3369) [inline test/runtests.jl]: Can we indent these lines?

File: test/runtests.jl [hunk 2]
```diff
@@ -173,6 +173,31 @@ module Interfaces
     flyto(::Crane, x::Float64, y::Float64) = "Arrvied at ($x, $y)"
     look_at_the_mirror_daily(::Crane) = true
 
+    struct Penguin end
+    @trait Dive
+    @assign Penguin with Dive
+    @implement CanDive by dive1(::Integer)           # missing argument name
```
Comment (tk3369) [inline test/runtests.jl]: My OCD kicks in :-) Can you change "missing" to "no" in the comment?  When I read missing, I got the impression that it's testing a negative case.

File: test/runtests.jl [hunk 3]
```diff
@@ -201,6 +226,11 @@ module Interfaces
             @test crane_check.implemented |> length == 3
             @test crane_check.misses |> length == 1
 
+            penguin_check = check(Penguin)
+            @test penguin_check.result == false
+            @test penguin_check.implemented |> length == (VERSION >= v"1.2" ? 7 : 4)
```
Comment (tk3369) [inline test/runtests.jl]: Since `VERSION >= v"1.2"` appears several times in this script, can we make a Bool at the top and use it everyone?

**Review by tk3369 — CHANGES_REQUESTED** [review]
It's a great attempt to make our own `has_method` function. I only have several minor requests.  Thanks

File: test/runtests.jl (hunk 1 above)
Comment (KlausC) [reply test/runtests.jl]: done

File: test/runtests.jl (hunk 2 above)
Comment (KlausC) [reply test/runtests.jl]: understood - correctness is a virtue if it does not exceed the limit of pedantic. I share your mindset.
done

File: test/runtests.jl (hunk 3 above)
Comment (KlausC) [reply test/runtests.jl]: done

**Comment by tk3369** [comment]
Great stuff!  Thanks for the quick turnaround. I'm pretty much ready to merge.  

My last check is test coverage.  I just managed to make the codecov report public. Could you add few more test cases for the new code according to the report?

File: test/runtests.jl [hunk 4]
```diff
@@ -194,12 +194,25 @@ module Interfaces
...
     dive5(::Penguin, ::Int) = 5                 # Int >: Bottom
     dive6(::Penguin, ::Int) = 6                 # not Int >: Number
 
+    # no contract requirements (code coverage)
+    struct Kiwi end
+
+    # weird argument types in contract specs 
+    @trait Creep
+    @implement CanCreep by creep1(a::Int=5) # argument assignment
+    @implement CanCreep by creep2({}) # invalid argument
```
Comment (tk3369) [inline test/runtests.jl:209]: Should probably raise an error instead of considering it as a `Bottom`? 

```
julia> @macroexpand(@implement CanCreep by creep({}))
quote
    function creep end
    BinaryTraits.register(CanCreep, creep, (Base.Bottom,), (), Any)
end
```

Comment (KlausC) [reply test/runtests.jl:209]: Ok, can do. Then I will accept only the following syntax:
```
<name> ::= <Symbol>
<argpart> ::= <name> OR <name> :: <type> OR :: <type>
<arg> ::= <argpart> OR <argpart> = <value>
<value> ::= <any Expr or Symbol>
```
and otherwise throw a Syntax exception.

Comment (KlausC) [reply test/runtests.jl:209]: As this PR is already closed, I will open a new one for these changes.
